common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3599,6 +3599,121 @@ static jint evp_aead_ctx_op(JNIEnv* env, jlong evpAeadRef, jbyteArray keyArray,
     return static_cast<jint>(actualOutLength);
 }
 
+static jint evp_aead_ctx_op_buf(JNIEnv* env, jlong evpAeadRef, jbyteArray keyArray, jint tagLen,
+                            jobject outBuffer, jbyteArray nonceArray,
+                            jobject inBuffer, jbyteArray aadArray,
+                            evp_aead_ctx_op_func realFunc) {
+
+    const EVP_AEAD* evpAead = reinterpret_cast<const EVP_AEAD*>(evpAeadRef);
+    JNI_TRACE("evp_aead_ctx_op(%p, %p, %d, %p, %p, %p, %p)", evpAead, keyArray, tagLen,
+              outBuffer, nonceArray, inBuffer, aadArray);
+
+     if (env->IsInstanceOf(inBuffer, conscrypt::jniutil::byteBufferClass) != JNI_TRUE || env->IsInstanceOf(outBuffer, conscrypt::jniutil::byteBufferClass) != JNI_TRUE  ) {
+        conscrypt::jniutil::throwException(env, "java/lang/IllegalArgumentException", "ByteBuffer Class Error");
+        return 0;
+    }
+
+    jboolean out_read = env->CallBooleanMethod(outBuffer,conscrypt::jniutil::byteBuffer_isReadMethod);
+    if(out_read == JNI_TRUE) {
+        conscrypt::jniutil::throwException(env, "java/nio/ReadOnlyBufferException", "output Read Only Buffer Exception");
+        return 0;
+    }
+
+    jboolean in_direct = env->CallBooleanMethod(inBuffer,conscrypt::jniutil::byteBuffer_isDirectMethod);
+    jboolean out_direct = env->CallBooleanMethod(outBuffer,conscrypt::jniutil::byteBuffer_isDirectMethod);
+    uint8_t* inBuf;
+    jint in_limit;
+    jint in_position;
+    jobject tmp;
+    if(in_direct == JNI_TRUE) {
+    // we are working with a direct ByteBuffer
+        inBuf = (uint8_t*)(env->GetDirectBufferAddress(inBuffer));
```

Comment:
This will return -1 unless `inBuffer` is a direct ByteBuffer, so you can drop the checks using JNI above and test the return value here.

Reply:
Ok I will change the logic around to remove thee JNI checks